rskj-core/src/main/java/org/ethereum/datasource/RocksDbDataSource.java

```diff
@@ -0,0 +1,364 @@
+                Set<ByteArrayWrapper> result = new HashSet<>();
+                for (iterator.seekToFirst(); iterator.isValid(); iterator.next()) {
+                    result.add(ByteUtil.wrap(iterator.key()));
+                }
+                if (logger.isTraceEnabled()) {
+                    logger.trace("<~ RocksDbDataSource.keys(): {}, {}", name, result.size());
+                }
+
+                return result;
+            }
+        } finally {
+            resetDbLock.readLock().unlock();
+            profiler.stop(metric);
+        }
+    }
+
+    private void updateBatchInternal(Map<ByteArrayWrapper, byte[]> rows, Set<ByteArrayWrapper> deleteKeys) {
+        Metric metric = profiler.start(Profiler.PROFILING_TYPE.DB_WRITE);
+        if (rows.containsKey(null) || rows.containsValue(null)) {
+            profiler.stop(metric);
+            throw new IllegalArgumentException("Cannot update null values");
+        }
+        // Note that this is not atomic.
+        try (WriteBatch batch = new WriteBatch()) {
+            for (Map.Entry<ByteArrayWrapper, byte[]> entry : rows.entrySet()) {
+                batch.put(entry.getKey().getData(), entry.getValue());
+            }
+            for (ByteArrayWrapper deleteKey : deleteKeys) {
+                batch.delete(deleteKey.getData());
+            }
+            db.write(new WriteOptions(), batch);
+            profiler.stop(metric);
+        } catch (RocksDBException e) {
+            logger.error("Exception. Not retrying.", e);
+        }
+
+    }
+
+    @Override
+    public void updateBatch(Map<ByteArrayWrapper, byte[]> rows, Set<ByteArrayWrapper> deleteKeys) {
+        if (rows.containsKey(null)) {
+            throw new IllegalArgumentException("Cannot update null values");
+        }
+        resetDbLock.readLock().lock();
+        try {
+            if (logger.isTraceEnabled()) {
+                logger.trace("~> RocksDbDataSource.updateBatch(): {}, {}", name, rows.size());
+            }
+
+            try {
+                updateBatchInternal(rows, deleteKeys);
+                if (logger.isTraceEnabled()) {
+                    logger.trace("<~ RocksDbDataSource.updateBatch(): {}, {}", name, rows.size());
+                }
+            } catch (IllegalArgumentException iae) {
+                throw iae;
+            } catch (Exception e) {
+                logger.error("Error, retrying one more time...", e);
+                // try one more time
+                try {
+                    updateBatchInternal(rows, deleteKeys);
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("<~ RocksDbDataSource.updateBatch(): {}, {}", name, rows.size());
+                    }
+                } catch (IllegalArgumentException iae) {
+                    throw iae;
+                } catch (Exception e1) {
+                    logger.error("Error", e);
+                    logger.error("Error", e1);
```

Comment:
Have two exceptions on the same place is bad, we need to rework that

Reply:
Done